Code review

nvgrw opened the pull request:
A pull request to allow project mentors to review the code

olpaw committed:
Fixes for native-image-agent build and restrict mode

- Add org.graalvm.compiler to AccessAdvisor internal class prefixes
- Fallback images built via Agent always obey specified restrictions

thurka committed:
[GR-16192] Backport to 19.0.1: GraalVM Version String Incorrect

nvgrw committed:
preliminary isomorphism and independence checks

nvgrw committed:
add rudimentary adjacency test

obtain correct type information to avoid hardcoding an offset suited
only to integer-size arrays

nvgrw committed:
type information incl primitive? in adj check

nvgrw committed:
NodePair => Pack, backed by AL

nvgrw committed:
pack combination function

nvgrw committed:
Wip implement follow use_defs

nvgrw committed:
implement prelim follow_def_uses

next: est_savings, alignment, investigate early bail conditions for
def_uses and use_defs.

nvgrw committed:
!inBlock => notInBlock

it's better to guard against not-in-block-ness rather than nesting
deeply. We don't want to operate on nodes that aren't in the current
block so the function has been inverted.

nvgrw committed:
label outer loops to avoid skipping next check on iterator

nvgrw committed:
refactor to remove state from arguments of utilities

nvgrw committed:
est_savings

next: alignment

nvgrw committed:
preliminary alignment support

need platform-specific and loop information to determine the vector
width to allow.

gilles-duboscq committed:
native-image: Remove redundant build_arg

(cherry picked from commit ee49a31bffc2e40148e1389e58c8261c11101957)

nvgrw committed:
tracing scheduling without dep cycle breaking

nvgrw committed:
remove B13 check

nvgrw committed:
remove packs with cyclic dependencies

nvgrw committed:
add vector value, read, write node

vector value may not be necessary

ansalond committed:
Remove forced mx version

ansalond committed:
[GR-15943] Fix the gates in the GraalVM 19.0 release branches.

PullRequest: graal/3678

olpaw committed:
[GR-16128] Backport to 19.0.1: Fix native-image-agent=restrict configuration detection.

nvgrw committed:
add vector support

nvgrw committed:
remove vector-specific types

nvgrw committed:
preliminary control flow w fixed vector nodes

nvgrw committed:
fix vector-like node verification methods

nvgrw committed:
fix VTS/STV canonicalization & cf order

ezzarghili committed:
[GR-15762] Backport to 19.0.1: Implement "build" mode for native-image-agent
[GR-15832] Backport to 19.0.1: Fallback image generation does not propagate --no-server

nvgrw committed:
use unrestricted stamps for STV

otherwise canonicalizer will reduce STV to (incorrect) constant node

ezzarghili committed:
[GR-15762] Backport to 19.0.1: Implement "build" mode for native-image-agent [GR-15832] Backport to 19.0.1: Fallback image generation does not propagate --no-server.

PullRequest: graal/3695

nvgrw committed:
rename VTS/STV to Pack/Unpack, fix canonicalization

nvgrw committed:
consider control flow in dependency analysis

nvgrw committed:
remove print

ezzarghili committed:
[GR-16128] Backport to 19.0.1: Fix native-image-agent=restrict configuration detection.

PullRequest: graal/3723

nvgrw committed:
hacky and wrong codegen

ezzarghili committed:
[GR-16176] Backport to 19.0.1: Ignore FallbackFeature when a FallbackExecutor gets built.

ezzarghili committed:
[GR-16176] Backport to 19.0.1: Ignore FallbackFeature when a FallbackExecutor gets built.

PullRequest: graal/3730

nvgrw committed:
change annotations for correct register allocation

ezzarghili committed:
[GR-15840] Backport to 19.0.1: Add testing of agent build mode.

PullRequest: graal/3713

ezzarghili committed:
[GR-16242] Backport to 19.0.1: Update FallbackExecutor API to accommodate JCK harness.

ezzarghili committed:
[GR-16242] Backport to 19.0.1: Update FallbackExecutor API to accommodate JCK harness.

PullRequest: graal/3733

ezzarghili committed:
[GR-16192] Backport to 19.0.1: GraalVM Version String Incorrect.

PullRequest: graal/3732

ezzarghili committed:
[GR-16219] Make jvmci-19-b03 release.

ezzarghili committed:
[GR-16219] Make jvmci-19-b03 release.

PullRequest: graal/3735

nvgrw committed:
add packing for constants on Intel ONLY

thomaswue committed:
Improve compilation performance of scheduler phase.

nvgrw committed:
fix some merge issues

still some issues remaining porting from 3 yo graal

cstancu committed:
Make FallbackExecutorRuntimeJavaArg a runtime option.

rschatz committed:
Skip source search path entries that we're not allowed to access.

nvgrw committed:
fix Isomorphic Packing after move to upstream

nvgrw committed:
add short names in effort to fix templating

nvgrw committed:
preliminary vector extract node

nvgrw committed:
remove VectorUnpackNode, add can for Extract

nvgrw committed:
untested vector extract codegen

nvgrw committed:
untested non-constant packing

cstancu committed:
[GR-16307] Backport to 19.0.1: Make FallbackExecutorRuntimeJavaArg a runtime option.

PullRequest: graal/3765

ezzarghili committed:
[GR-16330] Backport to 19.0.1: Improve compilation performance of scheduler phase.

PullRequest: graal/3766

ansalond committed:
[GR-15687] Deploy to the Nexus on postmerge.

(cherry picked from commit 663e3f530a7dbdefdfaeceef72410b7410c98e50)

ezzarghili committed:
[GR-16347] Backport to 19.0.1: Skip source search path entries that we're not allowed to access.

PullRequest: graal/3778

nvgrw committed:
preliminary vector integer stamp type

nvgrw committed:
emitVectorExtract => emitExtract

nvgrw committed:
remove toVectorKind

nvgrw committed:
remove isVector

nvgrw committed:
use vector stamps and generify some emit

nvgrw committed:
remove vectorAdd

nvgrw committed:
return Illegal rather than throw for stack kind

nvgrw committed:
wip replacement of vector add node

ansalond committed:
Adapt changes to the release brach.

ansalond committed:
[GR-16420] Backport: Deploy to the Nexus on postmerge.

PullRequest: graal/3798

ezzarghili committed:
[GR-16416] Make jvmci-19-b04 release.

ezzarghili committed:
[GR-16416] Make jvmci-19-b04 release.

PullRequest: graal/3801

ezzarghili committed:
[GR-16444] Release GraalVM 19.0.2.

nvgrw committed:
support arbitrary binary operations in vectorizer

nvgrw committed:
add codegen for + and -, 2/4 element doubleword

nvgrw committed:
move pack to separate file

nvgrw committed:
change alignments to integers

but I might've broken something. We will see.

nvgrw committed:
separate pairs and packs

nvgrw committed:
remove unused methods from ipp

nvgrw committed:
fix combine_packs broken after 555ce87

nvgrw committed:
add loop unrolling phase to high tier

currently appears to unroll x4

nvgrw committed:
move partial unroll to prevent double/excessive ur

nvgrw committed:
move unroll and always partially unroll loops to 4

nvgrw committed:
remove redundant getStackKind()

nvgrw committed:
clean up NodeView use

nvgrw committed:
add whitelist of vectorization candidate nodes

nvgrw committed:
stamp vector conversion modelled after constants

removes dependency on VectorIntegerStamp from places that shouldn't
depend on this

Commit:
Added VectorDescription, a new class to provide backend-specific vectorization info.

nvgrw committed:
pass checkstyle

Commit:
Updated a few things to match remote.

nvgrw committed:
IPP further checkstyle fixes

nvgrw committed:
further checkstyle fixes

nvgrw committed:
further formatting

nvgrw committed:
remove vectorWidth

nvgrw committed:
remove git merge conflict markers

nvgrw committed:
preliminary floating point support

Commit:
Started adding multiplication.

nvgrw committed:
add copyright notice to VLP

Commit:
Added AMD64 floating-point vector support.

Commit:
Made some functions static.

Commit:
Added support for vectorized multiplication and division to AMD64, along with changes to opcode selection.

nvgrw committed:
move VectorDescription and wire through LoopPolicies

nvgrw committed:
singleton vector descriptions

nvgrw committed:
move singleton due to consistency SPARC

nvgrw committed:
vector width in bytes rather than vector element

Commit:
Add Mul and Div to IPP node whitelist, change VectorDescription to return number of elements.

nvgrw committed:
revert HEAD

Commit:
fix maxVectorWidth

Commit:
Readded multiplication and division to IPP whitelist.

nvgrw committed:
use min vector width to unroll

nvgrw committed:
revert formatting changes

Commit:
Fixed typo.

nvgrw committed:
fix memory alignment analysis

!! This may not be generic enough !!
Arrays appear to have a 16-byte-ish header that is not accounted for in
the alignment calculations. With this change we can pack larger arrays.

nvgrw committed:
fix verification errors

run with -ea

Commit:
First pass at fixing extract instructions.

Commit:
First pass at fixing pack instructions.

Commit:
Changed insert/extract generation to comply with checkstyle.

nvgrw committed:
null check base pointers in adjacency check

Commit:
Moved insert instructions to new VexRRMIOp instruction type.

Commit:
Actually started using temporary scratch register in extract.

Commit:
Changed emitReverse to emit in insert nodes, this seems to be correct as it yields consistent behavior with C1.

nvgrw committed:
refactor pack into own class for reg allocation

nvgrw committed:
optimize imports

nvgrw committed:
refactor & fix vector extract

nvgrw committed:
extract with toRegisterKind

nvgrw committed:
filter out new nodes in dependency check

Commit:
Removed SignedDivNode from whitelist, integer division can't be vectorized on AMD64.

Commit:
Added VectorFloatStamp mul and div ops.

nvgrw committed:
fix successor of predecessor in write emission

Commit:
Added an iterator check. Seems to solve a NoSuchElementException.

nvgrw committed:
add further iterator check

Commit:
Experimental support for vectorized bitwise and prefix arithmetic.

nvgrw committed:
add some of the default unrolling checks to vlp

nvgrw committed:
remove templated names for vector read and write

nvgrw committed:
optimize imports

nvgrw committed:
checkstyle fixes for AMD64ArithmeticLIRGenerator

nvgrw committed:
switch from getStackKind to stamp and meta access

We used getStackKind to determine the size of a value (in bytes). As it
turns out, values that are smaller than 4 bytes are padded to be 4 bytes
on the stack. For our application it is not acceptable to treat values
this way since we need to determine properties such as adjacency
accurately.

This commit switches to a different method, which uses the javaType
method on a Stamp and the context's meta access instance. This seems to
give correct results.

The error manifested itself in the BytecodeParser and other parts of the
runtime being unable to load methods during the intrinsics replacement
phase. NoSuchMethodError ended up occurring because methods are looked
up in a hash map that uses the hashCode of two strings.
StringLatin1.hashCode was not compiled properly with vectorization and
exposed that the implementation was not capable of handling types that
are less than 4 bytes large (1 byte in the case of
StringLatin1.hashCode) correctly.

nvgrw committed:
optimize imports

nvgrw committed:
remove redundant code from 2f5757bcc315737dce6d795d3eb8225e4cc72fc9

nvgrw committed:
fix graph verification related error in vec write

also factor same array access check into a new static method

Commit:
Started unary arithmetic support in IPP.

Commit:
Merged proper loop policies, fixed up indentation.

Commit:
Bandaid removal of 2-byte vector operations - we don't support non-vector platform kinds and WORD was breaking stuff.

nvgrw committed:
clear successors in order to unregister properly

Commit:
Removed special behavior altogether for 2-byte vectors.

nvgrw committed:
handle duplicate packing nodes correctly

nvgrw committed:
fix unique usage in node inputs

nvgrw committed:
fix non-primitive access stamp alignment checks

nvgrw committed:
reg2stack/stack2reg for vector AMD64Kinds

this is untested and may not work. They're just MOV instructions.

Commit:
Made pack elements final.

Commit:
Added alternate packing operation, should be safer and potentially faster.

nvgrw committed:
result -> vector

nvgrw committed:
fix register checks for extract

nvgrw committed:
remove assertions that break narrow

Commit:
Started on a stack-based write instruction.

nvgrw committed:
fix array check and circular usage

nvgrw committed:
add check for ensuring block can be vectorized

Commit:
Renamed op, working on fixing value packing in the stack space.

nvgrw committed:
clean up codegen

nvgrw committed:
optimize imports in AMD64Unary

nvgrw committed:
remove STACK from packing temp, prevent vec2stack

nvgrw committed:
stamp for pack

nvgrw committed:
(temp) limit packs to platform vector size

Commit:
Added stack-based store op.

nvgrw committed:
stmtsCanPack now support null (TOP) alignment

I chose to pass an Integer that's nullable rather than Optional<Integer>
because of optional parameters being a discouraged practice:
https://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments

Commit:
Initialized temp in unary vector ops.

Commit:
Removed debug output.

nvgrw committed:
fix checkstyle errors

nvgrw committed:
reformat

nvgrw committed:
composite values alive

Commit:
Added movb address instruction.

nvgrw committed:
remove limiter

Commit:
Fixed constant packing op.

nvgrw committed:
update phase ordering to un-break IPP

nvgrw committed:
remove run script

nvgrw committed:
Merge tag 'vm-19.1.1' into feature/autovec

Commit:
Removed vector op table.

nvgrw committed:
checkstyle in AbstractPolyglotImpl

nvgrw committed:
fix checkstyle issues

Commit:
Slight optimization to packs - now properly combines two last bytes of array.

nvgrw committed:
change packing operation to use scalar type for mov

nvgrw committed:
add comment to improve packop for xmm-only devices

or maybe we should just generate those instructions regardless if we
don't have enough elements to fill a ymm-sized vector

nvgrw committed:
remove assertion in movb

nvgrw committed:
fix loop duplication invalidation bug

LoopTransformations duplicates loop fragments when inserting pre- and
post-loops. One would think that a method that duplicates a fragment
would not modify the original fragment, but the method present here does
just that. Rather than modifying the logic to do with duplication, we
have chosen to instead invalidate the loop's LoopFragmentInside in order
to ensure that both fragments are consistent with each other.

The processPreLoopPhis method processes the LoopFragmentInside and
traverses all of the nodes that are part of this fragment. It asserts
that there is a valid duplicate for each node that is present in the
inside fragment. However, this is not always the case as the duplication
removes some nodes from the LoopFragmentWhole rather than duplicating
them. Since these nodes aren't duplicated, the assertion fails.

nvgrw committed:
remove movb byte register assertion x2

nvgrw committed:
add a few verbose debug prints

Commit:
Added exception in unreachable case.

nvgrw committed:
workaround for single-vector blocks

Commit:
Added autovectorization toggle flag.

rramakrishna commented on compiler/src/org.graalvm.compiler.asm.amd64/src/org/graalvm/compiler/asm/amd64/AMD64Assembler.java:
```diff
@@ -949,10 +949,10 @@ public boolean check(AMD64 arch, int l, Register r, Register v, Register m, Regi
             }
             if (r != null) {
-                assert r.getRegisterCategory().equals(rCategory);
+                assert r.getRegisterCategory().equals(rCategory) : "r register " + r.name + " with category " + r.getRegisterCategory() + " does not match expected category " + rCategory;
```
Test, please ignore.

Commit:
Removed VexRRMIOp, was redundant as VexRVMIOp exists.

Commit:
Changed one-line if for checkstyle compliance.

nvgrw committed:
change Autovectorize flag to expert, unroll cond

nvgrw committed:
cleanup

nvgrw committed:
add space back to ComponentInstaller

Commit:
Decluttered AVXSize enum.

Commit:
Removed magic numbers from PackConstantsOp implementation.

Commit:
Documented twice() helper.

Commit:
Commented PackConstantsOp.

nvgrw committed:
store results from depth check

Commit:
Removed unused import.

Commit:
Commented and cleaned up LoadStackOp.

Commit:
Commented and cleaned up StoreStackOp.

Commit:
Commented and cleaned up PackStackOp.

Commit:
Commented and cleaned up PackOp.

Commit:
Made AVXSize bytes private.

nvgrw commented on compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java:
```diff
@@ -74,18 +83,29 @@ private static AMD64Kind twice(AMD64Kind kind) {
                 return AMD64Kind.QWORD;
             default:
-                return kind;
+                throw GraalError.shouldNotReachHere("Unable to double AMD64Kind '" + kind.toString() + "'");
```
A personal taste thing but I prefer format strings over string concatenation

nvgrw commented on compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java:
```diff
@@ -101,26 +121,62 @@ public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
                     break;
                 default:
-                    throw GraalError.shouldNotReachHere();
+                    throw GraalError.shouldNotReachHere("Unsupported constant size '" + AVXKind.AVXSize.fromBytes(byteBuffer.capacity()));
```
String concat

nvgrw commented on compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java:
```diff
@@ -101,26 +121,62 @@ public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
+            reg2addr(crb, masm, srcKind, dst, asRegister(src));
+        } else if (isStackSlot(src)) {
+            // We move first to the scratch register, then to our destination.
+            // Is there a faster solution?
```
If this is something that you wish to revisit, please mark it TODO

nvgrw commented on compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java:
```diff
@@ -101,26 +121,62 @@ public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
+     * @param dst The address we want to move to.
+     * @param src The value we want to move from. Can be a register, stack variable, or constant.
+     * @param srcKind The AMD64Kind of the src value.
+     * @param scratch A scratch register to use if necessary. Must be allocated by the caller as an @Temp.
```
nit: a `@Temp`

nvgrw commented on compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java:
```diff
@@ -190,27 +246,32 @@ private static void const2addr(CompilationResultBuilder crb, AMD64MacroAssembler
+        // Our destination should always be a register.
         @Def({REG}) private AllocatableValue result;
+
+        // We need a scratch register for any possible memory-to-memory moves (see move(...) above).
```
I think there's some way to link to other methods? Might be worth investigating the exact Javadoc syntax for this but they use it a bunch throughout the codebase

nvgrw commented on compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java:
```diff
@@ -220,68 +281,98 @@ protected LoadStackOp(LIRInstructionClass<? extends LoadStackOp> c, LIRGenerator
-                // Lowest multiple of YMM_LENGTH_IN_BYTES that can fit all elements.
-                int amt = (int) Math.ceil(((double) sizeInBytes) / YMM_LENGTH_IN_BYTES) * YMM_LENGTH_IN_BYTES;
+                // Allocate a register-width of stack space for us to dump our values to.
+                masm.subq(rsp, YMM.getBytes());
```
Why restrict to YMM?

nvgrw commented on compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java:
```diff
@@ -220,68 +281,98 @@ protected LoadStackOp(LIRInstructionClass<? extends LoadStackOp> c, LIRGenerator
-        private static final int YMM_LENGTH_IN_BYTES = 32;
         public static final LIRInstructionClass<StoreStackOp> TYPE = LIRInstructionClass.create(StoreStackOp.class);
 
+        // The array we're writing to.
```
Not necessarily an array?

nvgrw commented on compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java:
```diff
@@ -303,57 +397,82 @@ public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
+        // Our destination should always be a register.
         @Def({REG}) private AllocatableValue result;
+
+        // We need a scratch register for any possible memory-to-memory moves (see move(...) above).
```
Linking thing again

nvgrw commented on compiler/src/org.graalvm.compiler.asm.amd64/src/org/graalvm/compiler/asm/amd64/AVXKind.java:
```diff
@@ -41,26 +41,36 @@
-                    return 64;
+        DWORD(4),
+        QWORD(8),
+        XMM(16),
```
Are we sure that this isn't somewhere else in JVMCI/Graal? There are constants used for the AMD64 registers in various places (`xmm0`, etc) - is it worth repeating this?

nvgrw reviewed (commented):
just some comments to perhaps consider :D

nvgrw committed:
partial port of alignment analysis from c2

nvgrw committed:
add AVExclusions flag to skip IPP for debugging

Commit:
Adjusted addresses to take our changed stack pointer into account.

nvgrw committed:
fix exclusions in empty string case

nvgrw committed:
remove combined pack set size limit

nvgrw committed:
use allocator for on-stack packing

nvgrw committed:
remove duplication in AMD64Packing

nvgrw committed:
delegate for source/destination addresses

nvgrw committed:
remove unused StackOp constructor

nvgrw committed:
fix stackop combination

nvgrw committed:
remove comment about displacement helper

nvgrw committed:
utilize displacement helper throughout AMD64Packing

nvgrw committed:
rearrange AMD64Packing

nvgrw committed:
fix duplication issue by saving scratch to stack for mem2mem

Commit:
Added node cycles and size estimates to vector nodes.

Commit:
Fixed single-precision vectorized division.

nvgrw committed:
add preliminary iv checker

nvgrw committed:
print details of instructions being packed

nvgrw committed:
add ability to whitelist and blacklist AV methods

nvgrw committed:
fix double precision float extraction

nvgrw committed:
more granular logging for IPP

nvgrw committed:
vmovdqu rather than stack-pack for xmm-sized values

nvgrw committed:
add vmovdqu128 emit XMM-only stack pack correctly

nvgrw committed:
appease CheckGraalInvariants

nvgrw committed:
ensure that only values of the same type are packed

nvgrw committed:
don't vectorize across membar

nvgrw committed:
add else

nvgrw committed:
fix reg2stack, stack2reg for vectors

nvgrw committed:
move phases files into subpackage

nvgrw committed:
checkstyle fixes

nvgrw committed:
separate heuristic and graph structure methods

nvgrw committed:
rearrange BlockInfo and add copyright notice

nvgrw committed:
move validForBlock

next i'm going to move around a bunch of the methods in IPP, then I'll
probably reduce the use of Streams in line with the rest of Graal

nvgrw committed:
stub pack filtering in AVPolicies

nvgrw committed:
categorize methods

nvgrw committed:
take view & policies from IPP

nvgrw committed:
add documentation comments and remove unnecessary predicate

nvgrw committed:
comments in BlockInfo

nvgrw committed:
remove VectorSupport

nvgrw committed:
remove fromPackElements builders

nvgrw committed:
revert FloatStamp back to upstream version

so that this change has minimal impact

nvgrw committed:
remove unused length variable

nvgrw committed:
remove unused length variable

nvgrw committed:
fix single precision vector add, sub, mul

nvgrw committed:
add AutovectorizationContext to pass context info

nvgrw committed:
add naive filtering heuristic
